dds_web/api/schemas/token_schemas.py

```diff
@@ -0,0 +1,52 @@
+"""MFA related marshmallow schema."""
+
+####################################################################################################
+# IMPORTS ################################################################################ IMPORTS #
+####################################################################################################
+
+# Standard Library
+
+# Installed
+import marshmallow
+
+# Own modules
+from dds_web import auth
+
+
+####################################################################################################
+# SCHEMA ################################################################################## SCHEMA #
+####################################################################################################
+
+
+class TokenSchema(marshmallow.Schema):
+    """Schema for token authentication used when acquiring an encrypted JWT."""
+
+    # Authentication One-Time code
+    HOTP = marshmallow.fields.String(
+        required=False,
+        validate=marshmallow.validate.And(
+            marshmallow.validate.Length(min=8, max=8),
+            marshmallow.validate.ContainsOnly("0123456789"),
+        ),
+    )
+
+    class Meta:
+        unknown = marshmallow.EXCLUDE
+
+    @marshmallow.validates_schema(skip_on_field_errors=True)
+    def validate_mfa(self, data, **kwargs):
+        """Verify HOTP (authentication One-Time code) is correct."""
+
+        # This can be easily extended to require at least one MFA method
+        if "HOTP" not in data:
+            raise marshmallow.ValidationError("MFA method not supplied")
+
+        user = auth.current_user()
+        if "HOTP" in data:
+            value = data.get("HOTP")
+            # Raises authenticationerror if invalid
+            user.verify_HOTP(value.encode())
```

Comment:
I have a suggestion for line 48. What do you think? Or could you make it even better?
The idea is if the entered otp value has expired or a wrong value is input and it has been more than 15 mins since otp has been issued, we will check and send a new email. As with the error, I didn't want to handle this in the model, so caught it only to raise again after sending email, if at all.

```
try:
    user.verify_HOTP(value.encode())
except AuthenticationError as error:
    dds_web.dds_web.security.auth.send_hotp_email(user)
    raise error
```

Reply:
If you don't want this, it means that when a user's hotp verification fails and the hotp time expires, they have to do basic auth themselves again to be able to get a new hotp. This may not be obvious.

Reply:
I thought that's how it should work, at least for the CLI? But if not, could you write a list on how it should work for the web and cli? When someone needs to do basic auth and mfa.

Reply:
OK, so I had a look at how the new cli behaviour is and it can prompt the user with "Second factor authentication failed, would you like to try again?". So I have a new suggestion, instead of making a change here, let's take `send_hotp_email(user)` from line 105 to line 101 in auth.py. This way a new value will be generated if needed, the first try to verify hotp will fail (it will fail anyway because it has been more than 15 mins already), but at least they get the email as hotp verify fails and the cli will prompt to try again. How about this?

Reply:
Sounds good!